backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+                    // shared with filter model, so should work but has not been tested. Error
+                    // message is preserved for filter model because existing tests expect it.
+                    // TODO: handle width > 64 bits for filter model
+                    if (table->program->options.arch.isNullOrEmpty() ||
+                        table->program->options.arch == "filter") {
+                        ::error(ErrorType::ERR_UNSUPPORTED,
+                                "%1%: fields wider than 64 bits are not supported yet", fieldName);
+                    }
+                }
+            }
+
+            bool isLPMKeyBigEndian = false;
+            if (table->isLPMTable()) {
+                if (c->matchType->path->name.name == P4::P4CoreLibrary::instance().lpmMatch.name)
+                    isLPMKeyBigEndian = true;
+            }
+
+            builder->emitIndent();
+            if (memcpy) {
+                builder->appendFormat("__builtin_memcpy(&(%s.%s[0]), &(", keyname.c_str(),
+                                      fieldName.c_str());
+                table->codeGen->visit(c->expression);
+                builder->appendFormat("[0]), %d)", scalar->bytesRequired());
+            } else {
+                builder->appendFormat("%s.%s = ", keyname.c_str(), fieldName.c_str());
+                if (isLPMKeyBigEndian) builder->appendFormat("%s(", swap.c_str());
+                table->codeGen->visit(c->expression);
+                if (isLPMKeyBigEndian) builder->append(")");
+            }
+            builder->endOfStatement(true);
+
+            cstring msgStr, varStr;
+            if (memcpy) {
+                msgStr = Util::printf_format("Control: key %s", c->expression->toString());
+                builder->target->emitTraceMessage(builder, msgStr.c_str());
+            } else {
+                msgStr = Util::printf_format("Control: key %s=0x%%llx", c->expression->toString());
+                varStr = Util::printf_format("(unsigned long long) %s.%s", keyname.c_str(),
+                                             fieldName.c_str());
+                builder->target->emitTraceMessage(builder, msgStr.c_str(), 1, varStr.c_str());
+            }
+        }
+    }
+
+    builder->emitIndent();
+    builder->appendLine("struct p4tc_table_entry_act_bpf *act_bpf;");
+
+    builder->emitIndent();
+    builder->appendLine("/* value */");
+    builder->emitIndent();
+    cstring valueName = "value";
+    builder->appendFormat("struct %s *%s = NULL", table->valueTypeName.c_str(), valueName.c_str());
+    builder->endOfStatement(true);
+
+    if (table->keyGenerator != nullptr) {
+        builder->emitIndent();
+        builder->appendLine("/* perform lookup */");
+        builder->target->emitTraceMessage(builder, "Control: performing table lookup");
+        builder->emitIndent();
+        builder->appendLine("act_bpf = bpf_skb_p4tc_tbl_lookup(skb, &params, &key, sizeof(key));");
```

Comment:
nit: We usually emit table lookup helpers with `targetXXX->emitTableLookup()`. What about introducing a new `P4TCTarget` class and embedding this and other (future) helpers there?

Reply:
Functions from KernelSamplesTarget class is used mostly. table lookup call is emitted only one place and creating a new class and replicating all functions except for emitTableLookup() might lead to redundancy of code. If there are multiple lookup calls needed, may be in future, P4TCTarget class can be created by inheriting KernelSamplesTarget class

Reply:
ok - I'm fine with adding such a class once more kfuncs appear